src/feature/hs/hs_ob.c

```diff
@@ -218,3 +239,54 @@ hs_ob_parse_config_file(hs_service_config_t *config, const char *fname)
   tor_free(content);
   return ret;
 }
+
+/** Compute all possible subcredentials for every onion master key in the
+ * given service config object. The subcredentials is allocated and set as an
+ * continous array containing all possible values.
+ *
+ * The number of subcredential set in the array is returned (length). If the
+ * configuration has no OB master keys configured, 0 is returned and
+ * subcredentials is set to NULL. */
+size_t
+hs_ob_get_subcredentials(const hs_service_config_t *config,
+                         uint8_t **subcredentials)
+{
+  unsigned int num_pkeys, idx = 0;
+  int steps[3] = {0, -1, 1};
+  const uint64_t tp = hs_get_time_period_num(0);
+
+  tor_assert(config);
+
+  num_pkeys = smartlist_len(config->ob_master_pubkeys);
+  if (!num_pkeys) {
+    *subcredentials = NULL;
+    goto end;
+  }
+
+  /* Next loop will build 3 subcredentials for each public key. */
+  *subcredentials = tor_malloc_zero(DIGEST256_LEN * 3 * num_pkeys);
+
+  /* Time to build all the subcredentials for each time period: the previous
```

Comment:
Is this three-subcredentials thing only done for the onionbalance case?  If so, why?  If not, can we share the logic with whatever else does it?

Reply:
That's a good question. Let me try to answer it:

The reason we are not doing this brute force in the normal hsv3 case, is because the service always knows the right subcredential to use:

In `service_handle_introduce2()` the service gets the descriptor object that corresponds to this intro point, and because each descriptor corresponds to a specific subcredential, we get the right subcredential out of it, and use that to do the decryption. So that works just fine now that each descriptor has different intro points, however we will probably need to change this logic when we do #22893 (potentially by giving a different intro auth key to the same intro point in diff descriptors, so that we can still distinguish which descriptor it came off).